tinygrad/tensor.py

```diff
@@ -519,11 +519,11 @@ def log_softmax(self, axis=-1):
   def argmax(self, axis=None, keepdim=False):
     if axis is None:
       idx = (self == self.max(axis)) * Tensor.arange(prod(self.shape)-1,-1,-1, requires_grad=False, device=self.device).reshape(self.shape)
-      return prod(self.shape) - idx.max() - 1
```

Comment:
this is not the right fix, it's not `int` here due to implicit conversion in the ops, and casting at the end just hide the issue. it should automatically fixed with #2800

Reply:
I see, thanks for providing this context. However, I am not sure if #2800 will solve this. Multiplying 2 tensors above (line 521) will cast to `least_upper_dtype`. So if we take argmax of a Tensor with dtype `float32`, I don't see how this will result in a dtype of `int64`. 

Maybe I misunderstand #2800 though. If it does solve this issue, we can just keep the tests that I added in this PR and remove the explicit cast.

Reply:
the intended behavior is that `(self == self.max(axis))` would have `dtypes.bool`, `Tensor.arange` would have `dtypes.int32`, so multiplication output would be int32 (tinygrad's int default)

Reply:
Okay thank you, in that case is it okay that numpy's and tinygrad's dtypes differ? (`int64` vs `int32` respectively)

Reply:
jax argmax returns int32. int32 works better in GPU and accelerators. if we really want to use int64 then we will change the default type in arange, and that would propagate naturally. we don't need to manual cast here